In [29]:
import pandas as pd
import numpy as np

In [30]:
df = pd.read_csv('datos_inmuebles.csv')
df.head()

Tamaño_m2  Habitaciones  Ubicación  Año_Construcción  Precio_USD
0        152             5   Suburbio              2015    299781.0
1        229             5  Periferia              1991    385551.0
2        142             3   Suburbio              2020    258381.0
3         64             5  Periferia              1998    135449.0
4        156             4   Suburbio              1991    277521.0

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Tamaño_m2         200 non-null    int64  
 1   Habitaciones      200 non-null    int64  
 2   Ubicación         200 non-null    object 
 3   Año_Construcción  200 non-null    int64  
 4   Precio_USD        200 non-null    float64
dtypes: float64(1), int64(3), object(1)
memory usage: 7.9+ KB


In [32]:
df.describe()

Tamaño_m2  Habitaciones  Año_Construcción     Precio_USD
count  200.000000    200.000000         200.00000     200.000000
mean   174.995000      2.855000        1999.95000  306785.995000
std     75.042601      1.412071          12.14144  116055.509112
min     50.000000      1.000000        1980.00000   86941.000000
25%    106.250000      1.000000        1989.75000  208035.250000
50%    178.500000      3.000000        1999.50000  306324.500000
75%    241.250000      4.000000        2010.00000  416914.250000
max    299.000000      5.000000        2022.00000  524678.000000

In [33]:
df.isnull().sum()

Tamaño_m2           0
Habitaciones        0
Ubicación           0
Año_Construcción    0
Precio_USD          0
dtype: int64

In [34]:
df['Ubicación'].unique()

array(['Suburbio', 'Periferia', 'Centro'], dtype=object)

In [35]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['Ubicación'] = le.fit_transform(df['Ubicación'])
df

Tamaño_m2  Habitaciones  Ubicación  Año_Construcción  Precio_USD
0          152             5          2              2015    299781.0
1          229             5          1              1991    385551.0
2          142             3          2              2020    258381.0
3           64             5          1              1998    135449.0
4          156             4          2              1991    277521.0
..         ...           ...        ...               ...         ...
195        290             1          0              1994    488878.0
196        101             1          2              2016    182231.0
197        145             1          0              2000    259490.0
198        181             3          1              1993    290810.0
199        271             5          1              1981    438484.0

[200 rows x 5 columns]

In [36]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

X_train = df.drop('Precio_USD', axis=1)
y_train = df['Precio_USD']

np.random.seed(42)
X_test = np.random.normal(X_train.mean(), X_train.std(), size=(10, X_train.shape[1]))

model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)


mse = mean_squared_error(y_train, model.predict(X_train))
print('Error cuadrático medio:', mse)

Error cuadrático medio: 200079002.05442342


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [38]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

model_quad = PolynomialFeatures(degree=2)
X_train_quad = model_quad.fit_transform(X_train)

model_poly_reg = LinearRegression()
model_poly_reg.fit(X_train_quad, y_train)

y_pred_train_quad = model_poly_reg.predict(X_train_quad)

mse_quad = mean_squared_error(y_train, y_pred_train_quad)
print('Error cuadrático medio (regresión polinomial):', mse_quad)

Error cuadrático medio (regresión polinomial): 27913356.159237765


In [41]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error

model_tree = DecisionTreeRegressor()
model_tree.fit(X_train, y_train)

y_pred_train_tree = model_tree.predict(X_train)

mse_quad = mean_squared_error(y_train, y_pred_train_tree)
print('Error cuadrático medio (arboles de decision):', mse_quad)

Error cuadrático medio (arboles de decision): 0.0


Analicemos el Error Cuadrático Medio (MSE) para los tres modelos de regresión:

*   **Regresión Lineal:** El MSE fue aproximadamente 200,079,002. Esto indica un error relativamente alto, lo que sugiere que una relación lineal simple podría no capturar completamente los patrones en los datos.
*   **Regresión Polinomial (Grado 2):** El MSE fue aproximadamente 27,913,356. Este valor es significativamente menor que el MSE de la regresión lineal, lo que indica que una relación polinomial de grado 2 proporciona un ajuste mucho mejor a los datos de entrenamiento.
*   **Árbol de Decisión para Regresión:** El MSE en los datos de entrenamiento fue 0. Como se explicó anteriormente, un árbol de decisión puede lograr un error de entrenamiento cero al ajustarse perfectamente a todos los puntos de entrenamiento. Sin embargo, esto a menudo indica sobreajuste (overfitting) y es posible que el modelo no funcione bien con datos nuevos y no vistos.

**Comparación y Mejor Ajuste:**

Basado en el MSE en los datos de entrenamiento:

*   El **Árbol de Decisión para Regresión** tiene el MSE de entrenamiento más bajo (0), pero esto probablemente se deba a sobreajuste.
*   La **Regresión Polinomial (Grado 2)** tiene un MSE mucho menor que el modelo de Regresión Lineal, lo que sugiere que captura los patrones subyacentes en los datos de entrenamiento de manera más efectiva que el modelo lineal.

Si bien el Árbol de Decisión tiene una puntuación de entrenamiento perfecta, es crucial evaluar los modelos en datos no vistos (un conjunto de prueba) para determinar su verdadero rendimiento e identificar qué modelo generaliza mejor. Generalmente, se prefiere un modelo con un MSE de entrenamiento más alto pero un mejor rendimiento en un conjunto de prueba en lugar de un modelo sobreajustado.

Basándonos en los modelos entrenados y sus MSE en los datos de entrenamiento:

**Patrones observados en la predicción de precios:**

*   La **Regresión Lineal** mostró un MSE alto, lo que sugiere que una relación lineal simple no es suficiente para capturar la complejidad en la predicción de precios de inmuebles. Probablemente no está identificando adecuadamente cómo el tamaño, el número de habitaciones, la ubicación y el año de construcción interactúan para determinar el precio.
*   La **Regresión Polinomial (Grado 2)** tuvo un MSE significativamente menor, lo que indica que al considerar las interacciones y términos cuadrados de las características, el modelo puede ajustarse mucho mejor a los datos de entrenamiento. Esto sugiere que la relación entre las características y el precio no es puramente lineal y que hay efectos no lineales o interacciones importantes.
*   El **Árbol de Decisión para Regresión** obtuvo un MSE de 0 en los datos de entrenamiento. Esto implica que el modelo ha aprendido perfectamente los datos de entrenamiento, creando reglas de decisión complejas que segmentan el espacio de características para predecir el precio exacto de cada inmueble en el conjunto de entrenamiento.

**¿Es el modelo adecuado para la toma de decisiones en la compra o venta de propiedades?**

Basándonos únicamente en el rendimiento en los datos de entrenamiento:

*   La **Regresión Lineal** parece ser demasiado simple y no capturaría las complejidades del mercado inmobiliario para ser útil en la toma de decisiones importantes de compra o venta.
*   La **Regresión Polinomial (Grado 2)** muestra un mejor ajuste en los datos de entrenamiento, lo que la hace más prometedora que la regresión lineal. Podría ser un punto de partida para un modelo más robusto.
*   El **Árbol de Decisión para Regresión**, aunque tiene un MSE de 0 en el entrenamiento, probablemente está sobreajustado. Un modelo sobreajustado memoriza los datos de entrenamiento pero no generaliza bien a datos nuevos. Usar un modelo sobreajustado para la toma de decisiones en propiedades reales podría llevar a predicciones muy inexactas.

**Conclusión:**

Ninguno de los modelos, basándose solo en el MSE de entrenamiento, es directamente adecuado para tomar decisiones críticas de compra o venta de propiedades. El sobreajuste (en el caso del árbol de decisión) y el ajuste insuficiente (en el caso de la regresión lineal) son problemas importantes. La regresión polinomial muestra potencial, pero para que cualquier modelo sea confiable para la toma de decisiones en el mundo real, es **absolutamente necesario evaluarlo en un conjunto de datos de prueba independiente** para medir su capacidad de generalización y asegurarse de que no esté sobreajustado. Además, se necesitarían más datos y posiblemente otras características (como el estado del inmueble, servicios cercanos, etc.) para crear un modelo verdaderamente útil.

Aquí hay una explicación de cuándo sería recomendable utilizar cada tipo de modelo de regresión para problemas de predicción similares:

*   **Regresión Lineal:**
    *   **Cuándo usarla:** Cuando se sospecha que existe una relación lineal simple entre las variables predictoras y la variable objetivo. Es decir, cuando un cambio en una variable predictora resulta en un cambio constante en la variable objetivo. Es un buen punto de partida y es fácil de interpretar.
    *   **Ejemplos:** Predecir el salario en función de los años de experiencia (si la relación es aproximadamente lineal), predecir la demanda de un producto en función de su precio (si la relación es lineal).
    *   **Consideraciones:** Sensible a los valores atípicos (outliers). No captura relaciones no lineales complejas.

*   **Regresión Polinomial:**
    *   **Cuándo usarla:** Cuando la relación entre las variables predictoras y la variable objetivo es curvilínea o no lineal. Permite modelar curvas y relaciones más complejas que la regresión lineal.
    *   **Ejemplos:** Predecir el crecimiento de una planta en función del tiempo (que a menudo sigue una curva en forma de S), predecir el rendimiento de un cultivo en función de la cantidad de fertilizante aplicado (donde puede haber un punto óptimo).
    *   **Consideraciones:** Elegir el grado del polinomio adecuado es crucial. Un grado demasiado bajo puede no capturar la no linealidad, mientras que un grado demasiado alto puede llevar a sobreajuste. La interpretación puede ser menos directa que en la regresión lineal.

*   **Árboles de Decisión para Regresión:**
    *   **Cuándo usarlos:** Cuando las relaciones entre las características y la variable objetivo son complejas, no lineales y pueden involucrar interacciones entre características. Son buenos para capturar patrones complejos y no requieren suposiciones sobre la distribución de los datos.
    *   **Ejemplos:** Predecir el precio de una vivienda basándose en una variedad de características (tamaño, ubicación, número de habitaciones, etc.) donde las interacciones entre estas características son importantes, predecir la probabilidad de que un cliente compre un producto basado en su historial de navegación y datos demográficos.
    *   **Consideraciones:** Los árboles de decisión individuales pueden ser propensos al sobreajuste (como se vio en el ejemplo con MSE de 0). Para mejorar la generalización, a menudo se utilizan en conjuntos (ensembles) como Random Forests o Gradient Boosting Machines. Son menos interpretables que la regresión lineal para relaciones simples.